### PR TITLE
#1669 tradingview save template storage

### LINKED ISSUE #1669: [1.5] Tradingview save template storage
The trading view has the option to save a template, but last time the list was cleared and I had to redo the whole theme. 

Are there any plans how this is moderated in the future?

### PR DESCRIPTION
**Pull request by sschiessl-bcp**
<h2>General</h2>
Closes #1669

Enable saving and loading of presets from localStorage.

This issue required investigation in feasible login and storage methods and will be beneficial when adding others in the future, see https://github.com/blockchainprojects/bitshares-ui/issues/86 and corresponding PR mentioned in there for more details. This had a learning curve but brings much more insight in the possiblities.

Bounty requested: 7
Hours spent: 24.5

<h2>General</h2>
Please make sure the following is done:

- Pull request is onto develop
- If you haven't already, have a look at
  - https://github.com/bitshares/bitshares-ui/blob/develop/CODE_OF_CONDUCT.md
  - https://github.com/bitshares/bitshares-ui/blob/develop/CONTRIBUTING.md

<h2>Code Preparation</h2>

_Please review all your changes one last time before committing_

- [ ] Check for unused code
- [ ] No unrelated changes are included
- [ ] None of the changed files are reformatting only
- [ ] Code is self explanatory or documented
- [ ] All written text is properly translated (english language)

<h2>Testing</h2>

_The branch has been tested on the following browsers (desktop and mobile view)_

- [ ] Chrome 
- [ ] Opera
- [ ] Firefox
- [ ] Safari

<h2>User interface changes</h2>

_Delete this section if there weren't any UI changes. Please make sure you tested your changes in all themes_

- [ ] Dark
- [ ] Light
- [ ] Midnight

_Please provide screenshots/licecap of your changes below_

### CODE REVIEW

**Comment by sschiessl-bcp** [comment]
@VerevkinAlexander 

Please remove all reformatting only changes (push to blockchainprojects/bitshares-ui/86_tradingview_save_template_storage branch)

**Comment by startailcoon** [comment]
Nice work. Just a few issues when testing this out.

----

I get the following error on this (after removing the duplicate counterpart)
It solved itself after refreshing. Not sure what happened.

```
BindToChainState(ExchangeSubscriber) TypeError: "this.props.charts is undefined"
    render TradingViewPriceChart.jsx:286
```

* Can we fix so this overlay isn't in the way when the screen is less wide?

image

* Saving templates with identical names will overwrite old one. We should probably prompt the user about this, or disallow saving with a name that exists.

* The saved template uses a fixed quote/base pair. This means that if you load it from another pair it will show the saved pair and not the current pair. Either hide templates that doesn't match the current pair, or apply settings to the current pair.

image

* If the user loads a template from another pair like described before, then changes it and saves it, it will be stored as the new base pair in the list, while currency is still the old pair. 

* Removing a template will remove and close the modal. If the user wants to remove more than one it will have to open the modal several times. Don't close the modal on removing a stored template. 

* Switching from/to chart will not load the template used. Reproduce by 1. Selecting a template. 2. Switching to depth chart. 3. Switching back. Trading view will now reload without the selected template from before.

File: app/components/Exchange/TradingViewPriceChart.jsx [hunk 1]
```diff
@@ -128,21 +132,24 @@ class TradingViewPriceChart extends React.Component {
...
+                .append("<span>Load</span>");
             this.tvWidget
                 .createButton()
                 .attr("title", "Save Custom charts")
+                .addClass("apply-common-tooltip")
                 .on("click", () => {
                     that.setState({showSaveModal: true});
                 })
-                .append("<span>Save Chart</span>");
+                .append("<span>Save</span>");
```
Comment (startailcoon) [inline app/components/Exchange/TradingViewPriceChart.jsx]: Please utilize translation

File: app/components/Exchange/TradingViewPriceChart.jsx [hunk 2]
```diff
@@ -193,32 +202,67 @@ class TradingViewPriceChart extends React.Component {
...
     render() {
+        const {charts, quoteSymbol, baseSymbol} = this.props;
+        const {error} = this.state;
+        let _error = error ? "has-error" : "";
+        let dataSource = charts
+            .toArray()
+            .filter(chart => chart.symbol === quoteSymbol + "_" + baseSymbol);
         const columns = [
             {
                 title: "Layout Name",
```
Comment (startailcoon) [inline app/components/Exchange/TradingViewPriceChart.jsx]: Please utilize translation

File: app/components/Exchange/TradingViewPriceChart.jsx [hunk 3]
```diff
@@ -230,11 +274,6 @@ class TradingViewPriceChart extends React.Component {
                 dataIndex: "modified",
                 key: "modified"
             },
-            {
-                title: "Active Symbol",
-                dataIndex: "symbol",
-                key: "symbol"
-            },
             {
                 title: "Actions",
```
Comment (startailcoon) [inline app/components/Exchange/TradingViewPriceChart.jsx]: Please utilize translation

File: app/components/Exchange/TradingViewPriceChart.jsx [hunk 4]
```diff
@@ -305,11 +346,28 @@ class TradingViewPriceChart extends React.Component {
...
+                            </span>
+                        ) : null}
+                        <span
+                            className={_error}
+                            style={{
+                                borderBottom: "#A09F9F 1px dotted"
+                            }}
+                        >
+                            <Input
+                                placeholder="Enter Chart Layout Name"
```
Comment (startailcoon) [inline app/components/Exchange/TradingViewPriceChart.jsx]: Please utilize translation

File: app/components/Exchange/TradingViewPriceChart.jsx [hunk 5]
```diff
@@ -128,21 +132,24 @@ class TradingViewPriceChart extends React.Component {
             this.tvWidget
                 .createButton()
                 .attr("title", "Load custom charts")
+                .addClass("apply-common-tooltip")
                 .on("click", () => {
                     that.setState({showLoadModal: true});
                 })
-                .append("<span>Load Chart</span>");
+                .append("<span>Load</span>");
```
Comment (startailcoon) [inline app/components/Exchange/TradingViewPriceChart.jsx]: Please utilize translation

File: app/components/Exchange/TradingViewPriceChart.jsx [hunk 6]
```diff
@@ -128,21 +132,24 @@ class TradingViewPriceChart extends React.Component {
             this.tvWidget
                 .createButton()
                 .attr("title", "Load custom charts")
```
Comment (startailcoon) [inline app/components/Exchange/TradingViewPriceChart.jsx]: Please utilize translation

**Comment by sschiessl-bcp** [comment]
Thanks for the review startail

**Pull request merged**